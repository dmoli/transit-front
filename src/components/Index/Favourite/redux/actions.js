import { set as setRoute } from '../../Route/redux/actions';

/**
 * Action types
 */
export const actionTypes = {
  /** Set */
  SET_FAVOURITE: 'SET_FAVOURITE',
};

/**
 * Modify entities
 *
 * @param {array} entities entities
 * @return action to dispatch
 */
export const set = entities => ({
  type: actionTypes.SET_FAVOURITE,
  payload: entities,
});

/**
 * Find by id
 *
 * @param {int} routeId route id
 * @param {array} routes entities
 * @return entity found
 */
const findById = (routeId, routes) => {
  const setedRoutes = JSON.parse(JSON.stringify(routes));
  let entitySeted;
  for (const entity of setedRoutes) {
    if (routeId === entity.route_id) {
      entitySeted = entity;
      break;
    }
  }
  return entitySeted;
};

/**
 * Set object fields of the array by actions
 *
 * @param {int} id id of the entity
 * @param {array} entities entities
 * @param {string} action action to do
 * @return {array} setedEntities entities seted
 */
const setFields = (id, entities, action) => {
  const setedEntities = JSON.parse(JSON.stringify(entities));
  let i = 0;
  for (const entity of setedEntities) {
    if (id === entity.route_id) {
      const entitySeted = entity;
      if (action === 'fav') entitySeted.favorited = true;
      if (action === 'unfav') entitySeted.favorited = false;
      setedEntities[i] = entitySeted;
      break;
    }
    i += 1;
  }
  return setedEntities;
};

/**
 * Add favorite - Mock only
 *
 * @param {int} routeId route id
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const add = routeId => (
  async (dispatch, getState) => {
    const state = getState();
    const route = findById(routeId, state.routes.entities);
    route.favorited = true;
    const favourites = state.favourites.entities;
    favourites.push(route);
    // add this route to state
    dispatch(set(favourites));

    // refresh routes
    const routesSeted = setFields(routeId, state.routes.entities, 'fav');
    dispatch(setRoute(routesSeted));
  }
);

/**
 * Remove favorite - Mock only
 *
 * @param {int} routeId route id
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const remove = routeId => (
  async (dispatch, getState) => {
    const state = getState();
    let i = 0;
    const setedFavourites = JSON.parse(JSON.stringify(state.favourites.entities));
    // find route in favourite and remove it
    for (const entity of setedFavourites) {
      if (routeId === entity.route_id) {
        // delete setedFavourites[i];
        setedFavourites.splice(i, 1);
        break;
      }
      i += 1;
    }
    console.log('setedFavourites', setedFavourites);
    // refresh favourites
    dispatch(set(setedFavourites));

    // refresh routes
    const routesSeted = setFields(routeId, state.routes.entities, 'unfav');
    dispatch(setRoute(routesSeted));
  }
);
