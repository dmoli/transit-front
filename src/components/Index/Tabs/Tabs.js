/* eslint no-return-assign: [0] */
import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

/**
 * Tabs Component
 */
class Tabs extends Component {
  constructor() {
    super();

    this.handleTabClick = this.handleTabClick.bind(this);
  }

  /**
   * Notify the parent component that tabs has been changed
   *
  * @param tabName tab name
   */
  handleTabClick(tabName) {
    const { onChangeTab } = this.props;
    onChangeTab(tabName);
  }

  render() {
    const { selected } = this.props;

    return (
      <ContainerTabs>
        <Tab
          selected={selected === 'routes'}
          onClick={() => this.handleTabClick('routes')}
        >
          <FormattedMessage
            id='tabs.routes'
            defaultMessage='Recorridos'>
            {txt => (<TabName>{txt}</TabName>)}
          </FormattedMessage>
        </Tab>
        <Tab
          selected={selected === 'favourites'}
          onClick={() => this.handleTabClick('favourites')}
        >
          <FormattedMessage
            id='tabs.favourites'
            defaultMessage='Favoritos'>
            {txt => (<TabName>{txt}</TabName>)}
          </FormattedMessage>
        </Tab>
      </ContainerTabs>
    );
  }
}

Tabs.defaultProps = {
  selected: 'routes',
};

Tabs.propTypes = {
  /* Tab name selected */
  selected: PropTypes.string,
  /* function - Notify the parent component that tabs has been changed */
  onChangeTab: PropTypes.func.isRequired,
};

const ContainerTabs = styled.section`
  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 30px 0;
  width: 100%;
`;

const Tab = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px 40px;
  border-bottom: 3px solid transparent;
  :hover {
    cursor: pointer;
  }
  ${props => props.selected === true && `
    border-bottom: 3px solid #3aa0c8;
  `}
`;

const TabName = styled.span`
  display: flex;
`;

export default Tabs;
