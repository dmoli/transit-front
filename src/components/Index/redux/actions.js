import * as api from '../../../api/routes';

/**
 * Action types
 */
export const actionTypes = {
  /** Set .entities */
  SET_ROUTES: 'SET_ROUTES',
  /** Set .entities && .page */
  SET_NEXT_PAGE_ROUTES: 'SET_NEXT_PAGE_ROUTES',
};

/**
 * Modify entities
 *
 * @param {array} entities entities
 * @return action to dispatch
 */
export const set = entities => ({
  type: actionTypes.SET_ROUTES,
  payload: entities,
});

/**
 * Modify entities and pagination +1
 *
 * @param {array} entities results
 * @return action to dispatch
 */
export const setNextPage = entities => ({
  type: actionTypes.SET_NEXT_PAGE_ROUTES,
  payload: entities,
});

/**
 * Get entities - API
 *
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const get = () => (
  async (dispatch) => {
    try {
      // get response - API
      const response = await api.get();
      if (response.status !== 200) {
        throw new Error('error');
      }

      // response to json
      const data = await response.json();
      // dispatch action
      dispatch(set(data));
      // if not exist
      if (data.length === 0) {
        throw new Error('empty');
      }

      return data;
    } catch (e) {
      // error
      throw new Error(e.message);
    }
  }
);
