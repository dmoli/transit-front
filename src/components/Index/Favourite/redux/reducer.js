import { combineReducers } from 'redux';

import initialState from '../../../../redux/state';
import { actionTypes } from './actions';

/**
 * Entities reducer
 *
 * @param {array} state list of entities
 * @param action action to do
 */
const entitiesReducer = (state = initialState.favourites.entities, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_FAVOURITE:
      return action.payload;
    default:
      return state;
  }
};


const reducer = combineReducers({
  entities: entitiesReducer,
});

export default reducer;
