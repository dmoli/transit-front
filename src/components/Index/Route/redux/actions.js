/* eslint no-restricted-syntax: [0] */
import * as api from '../../../../api/routes';
import { set as setFavourite } from '../../Favourite/redux/actions';

/**
 * Action types
 */
export const actionTypes = {
  /** Set .entities */
  SET_ROUTES: 'SET_ROUTES',
  /** Set .entities && .page */
  SET_NEXT_PAGE_ROUTES: 'SET_NEXT_PAGE_ROUTES',
  /** Set .current */
  SET_CURRENT: 'SET_CURRENT',
  /** Set .shapes */
  SET_SHAPES: 'SET_SHAPES',
};

/**
 * Modify entities
 *
 * @param {array} entities entities
 * @param {int} page optional number page
 * @return action to dispatch
 */
export const set = (entities, page = 1) => ({
  type: actionTypes.SET_ROUTES,
  payload: {
    entities,
    page,
  },
});

/**
 * Modify entities and pagination +1
 *
 * @param {array} entities results
 * @return action to dispatch
 */
export const setNextPage = entities => ({
  type: actionTypes.SET_NEXT_PAGE_ROUTES,
  payload: entities,
});

/**
 * Modify current
 *
 * @param {string} currentName currentName
 * @return action to dispatch
 */
export const setCurrent = currentName => ({
  type: actionTypes.SET_CURRENT,
  payload: currentName,
});

/**
 * Modify shapes
 *
 * @param {array} shapes shapes
 * @return action to dispatch
 */
export const setShapes = shapes => ({
  type: actionTypes.SET_SHAPES,
  payload: shapes,
});

/**
 * Verify if a route exist in favourites
 *
 * @param {array} routes entities
 * @param {array} favourites entities
 * @return {array} entities seted
 */
const setFavoriteState = (routes, favourites) => {
  const setedRoutes = JSON.parse(JSON.stringify(routes));
  const setedFavourites = JSON.parse(JSON.stringify(favourites));
  let i = 0;
  for (const routeItem of setedRoutes) {
    for (const favouriteItem of setedFavourites) {
      if (routeItem.id === favouriteItem.id) {
        const routeSetedItem = routeItem;
        routeSetedItem.favorited = true;
        setedRoutes[i] = routeSetedItem;
      }
    }
    i += 1;
  }
  return setedRoutes;
};

/**
 * Set object fields of the array by actions
 *
 * @param {int} id id of the entity
 * @param {array} entities entities
 * @param {string} action action to do
 * @return {array} setedEntities entities seted
 */
const setFields = (id, entities, action) => {
  const setedEntities = JSON.parse(JSON.stringify(entities));
  let i = 0;
  if (action === 'current') {
    for (const entity of setedEntities) {
      const entitySeted = entity;
      if (id === entity.route_id) {
        entitySeted.current = true;
      } else {
        entitySeted.current = false;
      }
      setedEntities[i] = entitySeted;
      i += 1;
    }
  }

  i = 0;
  if (action === 'init_shape') {
    for (const entity of setedEntities) {
      const entitySeted = entity;
      entitySeted.lat = parseFloat(entitySeted.shape_pt_lat);
      entitySeted.lng = parseFloat(entitySeted.shape_pt_lon);
      setedEntities[i] = entitySeted;
      i += 1;
    }
  }
  return setedEntities;
};

/**
 * Refresh current route
 *
 * @param {int} routeId route id
 * @param {string} routeName route name
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const refreshCurrent = (routeId, routeName) => (
  async (dispatch, getState) => {
    const state = getState();
    // refresh routes
    const routesSeted = setFields(routeId, state.routes.entities, 'current');
    dispatch(set(routesSeted, state.routes.page));

    // refresh favourites
    const favouritesSeted = setFields(routeId, state.favourites.entities, 'current');
    dispatch(setFavourite(favouritesSeted));

    // refresh current field
    dispatch(setCurrent(routeName));
  }
);

/**
 * Get entities - API
 *
 * @param {string} text text to search
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const get = (text = '', page = null) => (
  async (dispatch, getState) => {
    try {
      const state = getState();
      const pageToSearch = page === null ? state.routes.page : page;
      // get response - API
      const response = await api.get(pageToSearch, text);
      if (response.status !== 200) {
        throw new Error('error');
      }

      // response to json
      const data = await response.json();
      // verify if a route exist in favourites
      const dataSeted = setFavoriteState(data, state.favourites.entities);
      // dispatch action
      dispatch(set(dataSeted));
      // if not exist
      if (dataSeted.length === 0) {
        throw new Error('empty');
      }

      return dataSeted;
    } catch (e) {
      // error
      throw new Error(e.message);
    }
  }
);

/**
 * Get next page of entities - API
 *
 * @param {string} text text to search
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const getNextPage = (text = '') => (
  async (dispatch, getState) => {
    try {
      const state = getState();
      // get response - API
      const response = await api.get(state.routes.page + 1, text);
      if (response.status !== 200) {
        throw new Error('error');
      }

      // response to json
      const data = await response.json();
      // if not exist
      if (data.length === 0) {
        throw new Error('empty');
      }

      // dispatch action
      dispatch(setNextPage(data));

      return data;
    } catch (e) {
      // error
      throw new Error(e.message);
    }
  }
);

/**
 * Get shapes - API
 *
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const getShapes = routeId => (
  async (dispatch) => {
    try {
      // get response - API
      const response = await api.getShapes(routeId);
      if (response.status !== 200) {
        throw new Error('error');
      }

      // response to json
      const data = await response.json();
      // if not exist
      if (data.length === 0) {
        throw new Error('empty');
      }

      // set init
      const dataSeted = setFields(null, data, 'init_shape');
      // dispatch action
      dispatch(setShapes(dataSeted));

      return data;
    } catch (e) {
      // error
      throw new Error(e.message);
    }
  }
);
