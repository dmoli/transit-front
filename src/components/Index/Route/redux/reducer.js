import { combineReducers } from 'redux';

import initialState from '../../../../redux/state';
import { actionTypes } from './actions';

/**
 * Entities reducer
 *
 * @param {array} state list of entities
 * @param action action to do
 */
const entitiesReducer = (state = initialState.routes.entities, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_ROUTES:
      return action.payload.entities;
    case actionTypes.SET_NEXT_PAGE_ROUTES:
      return state.concat(action.payload);
    default:
      return state;
  }
};

/**
 * Number page reducer
 *
 * @param {array} state number page
 * @param action plus +1 in the number page
 */
const pageReducer = (state = initialState.routes.page, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_ROUTES:
      return action.payload.page;
    case actionTypes.SET_NEXT_PAGE_ROUTES:
      return state + 1;
    default:
      return state;
  }
};

/**
 * Current name reducer
 *
 * @param {array} state current name
 * @param action change name
 */
const currentReducer = (state = initialState.routes.current, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_CURRENT:
      return action.payload;
    default:
      return state;
  }
};

/**
 * Shapes reducer
 *
 * @param {array} state list of shapes
 * @param action action to do
 */
const shapesReducer = (state = initialState.routes.shapes, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_SHAPES:
      return action.payload;
    default:
      return state;
  }
};


const reducer = combineReducers({
  entities: entitiesReducer,
  page: pageReducer,
  current: currentReducer,
  shapes: shapesReducer,
});

export default reducer;
