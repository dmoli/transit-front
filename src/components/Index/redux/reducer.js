import { combineReducers } from 'redux';

import initialState from '../../../redux/state';
import { actionTypes } from './actions';

/**
 * Entities reducer
 *
 * @param {array} state list of entities
 * @param action action to do
 */
const entitiesReducer = (state = initialState.routes.entities, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_ROUTES:
      return action.payload;
    case actionTypes.SET_NEXT_PAGE_ROUTES:
      return state.concat(action.payload);
    default:
      return state;
  }
};

/**
 * Number page reducer
 *
 * @param {array} state number page
 * @param action plus +1 in the number page
 */
const pageReducer = (state = initialState.routes.page, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_ROUTES:
      return 1;
    case actionTypes.SET_NEXT_PAGE_ROUTES:
      return state + 1;
    default:
      return state;
  }
};


const reducer = combineReducers({
  entities: entitiesReducer,
  page: pageReducer,
});

export default reducer;
