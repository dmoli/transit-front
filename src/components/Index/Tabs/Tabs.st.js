/* eslint import/no-extraneous-dependencies: [0] */
import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';
import { Provider } from 'react-redux';

import Tabs from './Tabs';
import { initStore } from '../../../redux/store';

const store = initStore();

storiesOf('Index/Tabs', module)
  .addDecorator(story => <Provider store={store}>{story()}</Provider>)
  .add('Tabs',
    withInfo('Tabs')(() =>
      <Tabs
        onChangeTab={action('onChangeTab')}
      />,
    ),
  )
  .add('Tabs - with favourite',
    withInfo('Tabs - with favourite')(() =>
      <Tabs
        onChangeTab={action('onChangeTab')}
        favouritesCount={19}
      />,
    ),
  );
