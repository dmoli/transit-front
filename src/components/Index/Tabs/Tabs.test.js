/* eslint-env jest */
import React from 'react';
import { Provider } from 'react-redux';

import Tabs from './Tabs';
import { mountWithIntl } from '../../../scripts/intl-enzyme-test-helper';
import { initStore } from '../../../redux/store';

const store = initStore();

describe('<Tabs />', () => {
  it('should rendered correctly', () => {
    const component = mountWithIntl(
      <Provider store={store}>
        <Tabs onChangeTab={() => {}} />
      </Provider>,
    );
    expect(component.html()).toMatchSnapshot();
  });
});
