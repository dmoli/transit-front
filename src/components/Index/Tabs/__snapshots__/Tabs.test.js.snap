// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> should rendered correctly 1`] = `"<section class=\\"Tabs__ContainerTabs-sc-165jd2c-0 Qcmcy\\"><div class=\\"Tabs__Tab-sc-165jd2c-1 eAXPFc\\"><span class=\\"green Tabs__TabName-sc-165jd2c-2 kRvwt\\">Recorridos</span></div><div class=\\"Tabs__Tab-sc-165jd2c-1 bfRbTp\\"><span class=\\"green Tabs__TabName-sc-165jd2c-2 kRvwt\\">Favoritos</span></div></section>"`;
